File: CrackingTheCodingInterview/Udemy/ReverseLinkedList2.js
const reverseBetween = function(head, m, n) {
    let currentPos = 1
    let currentNode = head;
    let start = head;
    while (currentPos < m) {
        start = currentNode;
        currentNode = currentNode.next;
        currentPos++;
    }

    let newList = null;
    let tail = currentNode;

    while (currentPos >= m && currentPos <= n) {
        let next = currentNode.next;
        currentNode.next = newList;
        newList = currentNode;
        currentNode = next;
        currentPos++;
    }

    start.next = newList;
    tail.next = currentNode; //we append it all together
    if (m == 1) {
        return newList
    }
    return head;
}

File: CrackingTheCodingInterview/Udemy/ReverseLinkedList2.java
package CrackingTheCodingInterview.Udemy;

public class ReverseLinkedList2 {
    public static void main(String[] args) {
        
    }

    public ListNode reverseBetween(ListNode head, int left, int right) {
        int currPos = 1;
        ListNode currNode = head;
        ListNode start = head;
        while (currPos != left) {
            start = currNode;
            currNode = currNode.next;
            currPos++;
        }
        
        ListNode prev = null;
        ListNode tail = currNode;
        
        while (currPos >= left && currPos <= right) {
            ListNode next = currNode.next;
            currNode.next = prev;
            prev = currNode;
            currNode = next;
            currPos++;
        }
        
        start.next = prev;
        tail.next = currNode;
        
        if (left == 1) return prev;
        
        return head;
        
        
    }
}